applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp

```diff
@@ -116,34 +126,57 @@ bool CoulombWithTensionCutOffImpl::IsAdmissibleSigmaTau(const Vector& rTrialSigm
 
 Vector CoulombWithTensionCutOffImpl::DoReturnMapping(const Properties& rProperties,
                                                      const Vector&     rTrialSigmaTau,
-                                                     CoulombYieldSurface::CoulombAveragingType AveragingType) const
+                                                     CoulombYieldSurface::CoulombAveragingType AveragingType)
 {
-    const auto apex = CalculateApex(mCoulombYieldSurface.GetFrictionAngleInRadians(),
-                                    mCoulombYieldSurface.GetCohesion());
+    Vector result = ZeroVector(2);
+
+    double kappa     = mEquivalentPlasticStrain;
+    double tolerance = 1.0;
+    while (tolerance > 1.0e-8) {
+        const auto apex = CalculateApex(mCoulombYieldSurface.GetFrictionAngleInRadians(),
+                                        mCoulombYieldSurface.GetCohesion());
+
+        const auto corner_point = CalculateCornerPoint(mCoulombYieldSurface.GetFrictionAngleInRadians(),
+                                                       mCoulombYieldSurface.GetCohesion(),
+                                                       mTensionCutOff.GetTensileStrength());
+
+        if (IsStressAtTensionApexReturnZone(rTrialSigmaTau, mTensionCutOff.GetTensileStrength(), apex)) {
+            return ReturnStressAtTensionApexReturnZone(mTensionCutOff.GetTensileStrength());
+        }
+
+        if (IsStressAtTensionCutoffReturnZone(rTrialSigmaTau, mTensionCutOff.GetTensileStrength(),
+                                              apex, corner_point)) {
+            return ReturnStressAtTensionCutoffReturnZone(
+                rTrialSigmaTau, mTensionCutOff.DerivativeOfFlowFunction(rTrialSigmaTau),
+                mTensionCutOff.GetTensileStrength());
+        }
+
+        if (IsStressAtCornerReturnZone(
+                rTrialSigmaTau, mCoulombYieldSurface.DerivativeOfFlowFunction(rTrialSigmaTau, AveragingType),
+                corner_point)) {
+            return corner_point;
+        }
+
+        // Regular failure region
+        result = ReturnStressAtRegularFailureZone(
+            rTrialSigmaTau, mCoulombYieldSurface.DerivativeOfFlowFunction(rTrialSigmaTau, AveragingType),
+            mCoulombYieldSurface.GetFrictionAngleInRadians(), mCoulombYieldSurface.GetCohesion());
 
-    if (IsStressAtTensionApexReturnZone(rTrialSigmaTau, mTensionCutOff.GetTensileStrength(), apex)) {
-        return ReturnStressAtTensionApexReturnZone(mTensionCutOff.GetTensileStrength());
-    }
+        double lambda = CalculatePlasticMultiplier(
+            rTrialSigmaTau, mCoulombYieldSurface.DerivativeOfFlowFunction(rTrialSigmaTau, AveragingType),
+            mCoulombYieldSurface.GetFrictionAngleInRadians(), mCoulombYieldSurface.GetCohesion());
+        double delta_kappa = CalculateEquivalentPlasticStrain(rTrialSigmaTau, AveragingType, lambda);
+        kappa += delta_kappa;
```

Comment:
You could simplify this a bit: 
```suggestion
        kappa += CalculateEquivalentPlasticStrain(rTrialSigmaTau, AveragingType, lambda);;
```